featurizers/fingerprints.py

```diff
@@ -4,65 +4,138 @@
 import numpy as np
 import pandas as pd
 import rdkit.Chem as Chem
-from rdkit.Chem.rdMolDescriptors import GetMorganFingerprintAsBitVect
 
 from base import FingerprintTransformer
 
+"""
+If during multiprocessing occurs MaybeEncodingError, first check if there isn't thrown any exception inside
+worker function! (That error isn't very informative and this tip might save you a lot of time)
+"""
 
-class MorganFingerprintAsBitVect(FingerprintTransformer):
+"""
+fp_descriptors needs to be here or inside _transform() of a specific class (cannot be defined inside the __init__() of
+that class), otherwise pickle gets angry:
+        TypeError: cannot pickle 'Boost.Python.function' object
+"""
```

Comment:
No tak, dlatego wysłałem ci kod z mojej magisterki:
```
def get_morgan_fingerprints(
    molecules: list[Mol], radius: int, bits_per_radius: int
) -> np.ndarray:
    n_bits = radius * bits_per_radius

    results = np.empty((len(molecules), n_bits), dtype=np.int8)

    def calc_morgan_fp(mol: Mol, radius: int, n_bits: int):
        from rdkit.Chem.rdMolDescriptors import GetMorganFingerprintAsBitVect

        return GetMorganFingerprintAsBitVect(mol, radius=radius, nBits=n_bits)

    with joblib.parallel_backend("loky", n_jobs=N_JOBS):
        fp_vals = joblib.Parallel()(
            joblib.delayed(calc_morgan_fp)(mol, radius=radius, n_bits=n_bits)
            for mol in molecules
        )

    for i, fp in enumerate(fp_vals):
        DataStructs.ConvertToNumpyArray(fp, results[i])

    return results
```

Każda klasa dziedzicząca jak będzie implementować `self._calculate_fingerprint(X)` (analogiczne do `calc_morgan_fp()` powyżej), to ominiecie ten błąd

Reply:
A tak w ogóle - zawrzyjcie to koniecznie w treści pracy inżynierskiej! To jest bardzo ciekawy i nieoczywisty problem w programowaniu równoległym, na poziomie serializacji obiektów. Tutaj dyskusja:
https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/CAD4fdRQ17uAaskd3hD59bY1fmkCye0OyKBpe0G9oCzBCKouVVQ%40mail.gmail.com/